casa/IO/MultiFile.cc

```diff
@@ -36,191 +37,752 @@
+    // Allocate a temp buffer if header too large or if O_DIRECT.
+    Bool hasRemainder = (totalSize > itsBlockSize);
+    MultiFileBuffer mfbuf(itsUseODirect || hasRemainder  ?  itsBlockSize:0,
+                          itsUseODirect);
+    char* iobuf = mfbuf.data();
+    // First write the remainder and adjust the header as needed.
+    if (hasRemainder) {
+      writeRemainder (mio, cio, iobuf);
+    } else {
+      // No remainder, so write the continuation blocknrs here.
+      // None are used in the current set.
+      writeVector (cio, itsHdrCont[0].blockNrs);
+      writeVector (cio, itsHdrCont[1].blockNrs);
+      itsNrContUsed[itsHdrContInx] = 0;
+      cio.write (2, itsNrContUsed);
+    }
+    // Writing the first part of the header is done as the last step.
     uChar* buf = const_cast<uChar*>(mio.getBuffer());
-    CanonicalConversion::fromLocal (buf, todo);       // header size
-    // Write the first part of the buffer at the beginning of the file.
-    // Use an aligned buffer for possible O_DIRECT alignment.
-    itsIO.seek (0);
+    todo = mio.length();
+    CanonicalConversion::fromLocal (buf+24, todo);  // header size
+    if (itsUseCRC) {
+      uInt crc = calcCRC (buf, todo);
+      crc = calcCRC (buf, todo);
+      CanonicalConversion::fromLocal (buf+20, crc);
+    }
+    if (itsTestMode > 0  &&  itsNrBlock > itsTestMode) {
+      // Mimic an abort by not writing the first header block.
+      cout << "MultiFile::writeHeader ended prematurely for nrBlock="
+           << itsNrBlock << endl;
+      itsTestMode = -2;    // indicate no more writing
+      return False;
+    }
     if (itsUseODirect) {
-      MultiFileBuffer mfbuf(itsBlockSize, itsUseODirect);
-      memcpy (mfbuf.data, buf, itsBlockSize);
-      itsIO.pwrite (itsBlockSize, 0, mfbuf.data);
+      memcpy (iobuf, buf, itsBlockSize);
+      itsIO->pwrite (itsBlockSize, 0, iobuf);
     } else {
-      itsIO.pwrite (itsBlockSize, 0, buf);
+      itsIO->pwrite (itsBlockSize, 0, buf);
+    }
+    return True;
+  }
+
+  void MultiFile::writeRemainder (MemoryIO& mio, CanonicalIO& cio,
+                                  char* iobuf)
+  {
+    // Get the size of the remainder.
+    Int64 todo = mio.length() - itsBlockSize;
+    // Use the other continuation set.
+    int newContInx = 1 - itsHdrContInx;
+    // Determine the nr of continuation blocks to be used.
+    // The first 8 bytes per block are used to point to the next one.
+    // Furthermore, the old and new cont-block vector have to be counted
+    // as well (length and all entries).
+    // So it can happen that an extra block is needed which in its turn
+    // might require an extra block, etc.
+    // To avoid that the free list is changed and needs to be rewritten,
+    // new continuation blocks are created by adding a new block.
+    // Obsolete continuation blocks are lost.
+    // Maybe change this by storing actual nr of cont blocks used!!!!!
```

Comment:
What is the status of this comment?

Reply:
I have to think about it. It's quite a long time ago that I made these changes.